# Analyzing Used Car Listings on eBay Kleinanzeigen

We will analyse a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle with a sample of 50,000 data points, but this version is a less-cleaned csv file version of original data.

The data dictionary is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing.
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean the data and analyze the included used car listings.

We will first import the Pandas and NumPy libraries.

In [1]:
import pandas as pd
import numpy as np

Now we will read the csv file into pandas and assign it to a variable called `autos`, as well as get some basic information.

In [2]:
autos = pd.read_csv('autos.csv', encoding ='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our data consists of 20 columns and 5000 entires, each one for a different car. Most of the columns show string values, and columns 
`vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` include 5-10% of null values.

# Clean Columns

We will first clean the name of the columns by doing the following changes:
* Change the columns from camelcase to snakecase.
* Change a few wordings to more accurately describe the columns.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

# Initial Data Exploration and Cleaning

We will start by exploring the data to find columns where we can clean the data.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Our initial observations are:

* There are two text columns where almost all of the values are the same, `seller` and`offer_type`.
* The `num_photos` column looks odd, so we need to check that.
* There are two columns, `price` and `auto`, which are numeric values with extra characters being stored as text.

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [7]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$2,459        1
$13,560       1
$2,569        1
$2,245        1
$40,400       1
Name: price, Length: 2357, dtype: int64

In [8]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

We will first drop the column `num_photos` where all the values are 0 and does not provide us with any useful information.

In [9]:
autos = autos.drop('num_photos', axis=1)

Then, we will clean the columns `price` and `odometer` and convert them to integer columns, as well as change the name to `odometer_km`.

In [10]:
autos['price'] = (autos['price']
                            .str.replace('$', '')
                            .str.replace(',', '')
                            .astype(int)
                            )
            
autos['odometer'] = (autos['odometer']
                            .str.replace('km', '')
                            .str.replace(',', '')
                            .astype(int)
                            )

autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

# Exploring Price

We will explore in more detail the price column.

In [11]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().head(20))
print(autos['price'].value_counts().sort_index(ascending=False).head(20))
print(autos['price'].value_counts().sort_index(ascending=True).head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       

We can do the following observations:

* The prices in this column seem rounded and many values coincide, going from 5,000 values to only 2,357 unique values, so it is possible that the people in this site tend to round the price values.
* We found out that 1,421 rows show a value of \$0 (around 2\% of all the cars), so we will remove them.
* The price values higher than \$350,000 seem to be unrealistic, so we will also take them out.
*  There are also some cars valued as low as \$1, but given that eBay is an auction site, there could legitimately be items where the opening bid is that low, so we will keep them.

In [12]:
autos = autos[autos['price'].between(1, 350000)]
print(autos["price"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


# Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

* `date_crawled`: added by the crawler (string)
* `last_seen`: added by the crawler (string)
* `ad_created`: from the website (string)
* `registration_month`: from the website (int)
* `registration_year`: from the website (int)

Let's explore the first three columns to check what kind of strings are they.

In [13]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Looks like all the values contain the date in the first part of the text. Therefore, we will study the distribution of the values using only the date part.

In [14]:
(autos['date_crawled']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

It appears that the site was crawled daily over roughly a one month period in March and April 2016, and the distribution of listings crawled on each day is uniform.

In [15]:
(autos['ad_created']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

Most of the ads crawled were created during the crawling period, however there are some dates as old as 9 months before.

In [16]:
(autos['last_seen']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing in the `last_seen` column, which allows us to determine when a car was sold.

However, in the last three days there is a disproportionate amount of 'last seen' values, which seem to see connected to the fact that the crawling ended at that time, so the realibility of the values seems low.

In [17]:
autos['registration_month'].describe()
(autos['registration_month']
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

0     0.092248
1     0.066282
2     0.060476
3     0.103017
4     0.083105
5     0.083002
6     0.087944
7     0.079419
8     0.064367
9     0.068568
10    0.073880
11    0.068218
12    0.069474
Name: registration_month, dtype: float64

In [18]:
autos['registration_year'].describe()
(autos['registration_year']
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64

The `registration_year` indicated indicates the age of the car. Looking at the column, we note some odd values. The minimum value is 1000 and the maximum is 9999.

# Dealing with Incorrect Registration Year Data

Because a car cannot be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. In addition, a registration year before 1900 seems also unrealistic. Let's calculate first how big is the percentage of the listings that fall outside the 1900-2016 in order to verify if it is safe to delete that data.

In [19]:
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Since only 3% are outside the 1900-2016 interval, we consider it is safe to remove them. We will also check how the new distribution of the `registration_year` column changed.

In [20]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that the majority of the listed cars were registered in the last 20 years.

# Exploring Price by Brand

Now, we will explore the price of the listed cars depending on their brand. First, we will take only the brands that appear frequently enough to obtain relevant results.

In [21]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

It appears that German manufacturers like Volkswagen and BMW represent almost 50% of the overall listings. Volkswagen is by far the most popular brand with more than 20%.

There are lots of brands that do not have a significant percentage of listings, so we will limit our analysis of the mean price to only brands representing more than 5% of total listings to obtain relevant results.

In [22]:
brand_counts = autos['brand'].value_counts(normalize=True)
popular_brands = brand_counts[brand_counts > 0.05].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [23]:
brand_mean_prices = {}
for brand in popular_brands:
    brand_prices = autos[autos['brand']==brand]
    mean_price = brand_prices['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

After taking only the brands that make up more than 5% of the listings, We can observe an important price gap between the mean price of them:

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive, less than half the price of the expensive ones.

Volkswagen is in between, which may explain its popularity according to the number of listings, since it may be a good inbetween option.

# Exploring Mileage

Finally, we will analyse the mean values for mileage according to the brand to see if there is come correlation using the columnd `odometer`.

In [25]:
brand_mean_mileages = {}
for brand in popular_brands:
    brand_prices = autos[autos['brand']==brand]
    mean_mileage = brand_prices['odometer_km'].mean()
    brand_mean_mileages[brand] = int(mean_mileage)

brand_mean_mileages

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In order to compare easily with the mean prices we calculated in the previous section, we will create a dataframe using as columns both the mean price and mean mileage.

In [32]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
print(bmp_series)

bmm_series = pd.Series(brand_mean_mileages).sort_values(ascending=False)
print(bmm_series)

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64
bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
ford             124266
dtype: int64


In [36]:
brand_info = pd.DataFrame(bmp_series, columns=['mean_price'])
brand_info['mean_mileage'] = bmm_series
brand_info

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

The values of car mileages do not vary as much as the prices do by brand, being all of them on the 120,000-135,000 km range, although we can observe a slight trend to the more expensive vehicles having higher mileage, which we can assume it is due to buyers wanting to make the most out of the investment they made, althought the difference is smaller than we could have expected.